common/types/list.go

```diff
@@ -97,16 +97,30 @@ func NewJSONList(adapter ref.TypeAdapter, l *structpb.ListValue) traits.Lister {
 
 // NewMutableList creates a new mutable list whose internal state can be modified.
 //
-// The mutable list only handles `Add` calls correctly as it is intended only for use within
-// comprehension loops which generate an immutable result upon completion.
-func NewMutableList(adapter ref.TypeAdapter) traits.Lister {
+// The mutable list only handles `Add` and `Type` calls correctly as it is intended
+// only for use within comprehension loops which generate an immutable result upon
+// completion.
+//
+// The returned value's ToImmutableList must be called to return a list for use in
+// CEL expressions.
+func NewMutableList(adapter ref.TypeAdapter) traits.MutableLister {
 	return &mutableList{
 		TypeAdapter:   adapter,
-		baseList:      nil,
 		mutableValues: []ref.Val{},
 	}
 }
 
+func (*mutableList) Contains(ref.Val) ref.Val { panic("invalid use of mutable list") }
```

Comment:
In general we try not to panic. There are ways to make these calls work as well. For example, a naive way would be to first convert `ToImmutableList` and then invoke the appropriate method. I think that would be preferred (and would fix the prior behavior of accessing `nil` memory), even if it's inefficient.  Thoughts?

Reply:
Yeah, I guess that's a design decision.

When I was originally looking at it I saw the possibility of automatically immutifying the list, so for example this would become `func (l *mutableList) Contains(val ref.Val) ref.Val { l.ToImmutableList().Contains(val) }`. This obviously become very inefficient if used, but is reasonable. I also think it's reasonable though to return an error to the user if they use a function in a way that is explicitly documented to not work as this code currently does; essentially we are mimicking a type system via that mechanism in the same way that reflect does.

I'm happy to go either way, though tend to lean toward the harsh prevention of people doing the wrong thing.

Reply:
When I'm writing an application, I agree with you about erroring as soon as possible. However, as a common library author it's usually much better to degrade gracefully than to cause a failure in the application that uses it. I'll admit, I never expected someone outside the CEL team to use this function when I wrote it. Oops :)

I think, perhaps, the following would be just as good, and a better user experience if someone were to use the `MutableList` in a way other than what I had originally intended.

```go
// a/common/types/list.go
func NewMutableList(adapter ref.TypeAdapter) traits.Lister {
       var mutableValues []ref.Val
        return &mutableList{
               baseList: &baseList{
                       TypeAdapter: adapter,
                       value:       mutableValues,
                       size:        0,
                       get:         func(i int) interface{} { return mutableValues[i] },
               },
               mutableValues: mutableValues,
        }
 }
 
 // mutableList aggregates values into its internal storage. For use with internal CEL variables only.
 type mutableList struct {
        *baseList
        mutableValues []ref.Val
 }

func (l *mutableList) Add(other ref.Val) ref.Val {
        otherList, ok := other.(traits.Lister)
	if !ok {
		return MaybeNoSuchOverloadErr(otherList)
	}
        for i := IntZero; i < otherList.Size().(Int); i++ {
                l.size++
                l.mutableValues = append(l.mutableValues, otherList.Get(i))
        }
        return l
```

Reply:
@kortschak I think the small example I provided might actually do everything that you would expect a `traits.Lister` to do.

Reply:
Yes, you're absolutely right!

Reply:
The warnings also go away. I have added a minor optimisation for the case where the other list is also a `*mutableList` making use of the runtime's awareness of size classes.